File: sdks/go/retry.go
package strellerminds

import (
	"errors"
	"math"
	"math/rand"
	"strings"
	"time"
)

type RetryOptions struct {
	Retries       int
	InitialDelay  time.Duration
	MaxDelay      time.Duration
	Multiplier    float64
	Jitter        bool
	OnRetry       func(err error, attempt int, delay time.Duration)
	IsRetryable   func(err error) bool
}

func defaultRetryOptions() RetryOptions {
	return RetryOptions{
		Retries:      5,
		InitialDelay: 200 * time.Millisecond,
		MaxDelay:     10 * time.Second,
		Multiplier:   2.0,
		Jitter:       true,
	}
}

func Retry[T any](task func() (T, error), opts *RetryOptions) (T, error) {
	var zero T
	options := defaultRetryOptions()
	if opts != nil {
		options = *opts
	}
	attempt := 0
	delay := options.InitialDelay
	var lastErr error

	for attempt <= options.Retries {
		res, err := task()
		if err == nil {
			return res, nil
		}
		lastErr = err
		retryable := true
		if options.IsRetryable != nil {
			retryable = options.IsRetryable(err)
		}
		if !retryable || attempt == options.Retries {
			return zero, err
		}
		computed := delay
		if options.Jitter {
			j := rand.Float64() + 1.0 // [1.0, 2.0)
			computed = time.Duration(math.Min(float64(options.MaxDelay), float64(delay)*j))
		}
		if options.OnRetry != nil {
			func() {
				defer func() { _ = recover() }()
				options.OnRetry(err, attempt+1, computed)
			}()
		}
		time.Sleep(computed)
		delay = time.Duration(math.Min(float64(options.MaxDelay), math.Max(1, float64(delay)*options.Multiplier)))
		attempt++
	}
	if lastErr != nil {
		return zero, lastErr
	}
	return zero, errors.New("retry failed with unknown error")
}

func IsTransientNetworkError(err error) bool {
	if err == nil {
		return false
	}
	msg := strings.ToLower(err.Error())
	substrings := []string{
		"timeout",
		"network",
		"temporary",
		"connection reset",
		"econnreset",
		"econnrefused",
		"503",
		"502",
		"429",
		"rate limit",
		"not_found",
	}
	for _, s := range substrings {
		if strings.Contains(msg, s) {
			return true
		}
	}
	return false
}

File: sdks/typescript/src/retry.ts
export type RetryOptions = {
    retries?: number;
    initialDelayMs?: number;
    maxDelayMs?: number;
    multiplier?: number;
    jitter?: boolean;
    onRetry?: (error: unknown, attempt: number, delayMs: number) => void;
    isRetryable?: (error: unknown) => boolean;
    signal?: AbortSignal;
};

function sleep(delayMs: number, signal?: AbortSignal): Promise<void> {
    return new Promise((resolve, reject) => {
        const timer = setTimeout(resolve, delayMs);
        if (signal) {
            const onAbort = () => {
                clearTimeout(timer);
                reject(new Error('Retry aborted'));
            };
            if (signal.aborted) {
                onAbort();
                return;
            }
            signal.addEventListener('abort', onAbort, { once: true });
        }
    });
}

export async function retry<T>(
    task: () => Promise<T>,
    options: RetryOptions = {}
): Promise<T> {
    const {
        retries = 5,
        initialDelayMs = 200,
        maxDelayMs = 10_000,
        multiplier = 2,
        jitter = true,
        onRetry,
        isRetryable,
        signal,
    } = options;

    let attempt = 0;
    let delayMs = Math.max(0, initialDelayMs);
    let lastError: unknown;

    while (attempt <= retries) {
        try {
            return await task();
        } catch (error) {
            lastError = error;
            const retryable = isRetryable ? isRetryable(error) : true;
            if (!retryable || attempt === retries) {
                throw error;
            }
            const computedDelay = jitter
                ? Math.min(maxDelayMs, Math.floor(delayMs * (1 + Math.random())))
                : Math.min(maxDelayMs, delayMs);
            if (onRetry) {
                try {
                    onRetry(error, attempt + 1, computedDelay);
                } catch {
                    // ignore onRetry errors
                }
            }
            await sleep(computedDelay, signal);
            delayMs = Math.min(maxDelayMs, Math.max(1, Math.floor(delayMs * multiplier)));
            attempt += 1;
        }
    }
    // Should never reach here
    throw lastError ?? new Error('Retry failed with unknown error');
}

export function isTransientNetworkError(error: unknown): boolean {
    if (!error) return false;
    const message = String((error as any).message ?? error).toLowerCase();
    // Heuristics for transient conditions
    return (
        message.includes('timeout') ||
        message.includes('timed out') ||
        message.includes('network') ||
        message.includes('temporary') ||
        message.includes('connection reset') ||
        message.includes('econnreset') ||
        message.includes('econnrefused') ||
        message.includes('503') ||
        message.includes('502') ||
        message.includes('429') ||
        message.includes('rate limit') ||
        message.includes('not_found') // polling window race
    );
}

